rsl_tools.py

```diff
@@ -0,0 +1,78 @@
+import requests
+import zipfile
+import shutil
+import os
+import json
+import subprocess
+from version import randomizer_commit, randomizer_version
+
+def download_randomizer():
+    zippath = 'randomizer.zip'
+
+    # Make sure an old zip isn't sitting around
+    if os.path.isfile(zippath):
+        os.remove(zippath)
+
+    # Download the zipped randomizer
+    r = requests.get(f'https://github.com/Roman971/OoT-Randomizer/archive/{randomizer_commit}.zip', stream=True)
+    with open(zippath, 'wb') as fp:
+        for chunk in r.iter_content():
+            fp.write(chunk)
+    
+    # Extract the zip file and add __init__.py
+    zf = zipfile.ZipFile(zippath)
+    zf.extractall('.')
+    os.rename(f'OoT-Randomizer-{randomizer_commit}', 'randomizer')
+    with open(os.path.join('randomizer', '__init__.py'), 'w') as fp:
+        pass
+
+    # Delete the zip file
+    os.remove(zippath)
+
+def check_version():
+    if os.path.isfile(os.path.join('randomizer', 'version.py')):
+        from randomizer import version as ootrversion
+        if ootrversion.__version__ == randomizer_version:
+            return
+        else:
+            print("Updating the randomizer...")
+            shutil.rmtree('randomizer')
+            download_randomizer()
+    else:
+        print("Downloading the randomizer...")
+        download_randomizer()
+    return
+
+
+# This function will take things from the GUI eventually.
+def init_randomizer_settings(worldcount=1):
+    rootdir = os.getcwd()
+    randomizer_settings = {
+        "rom": os.path.join(rootdir, 'data','ZOOTDEC.z64'),
+        "output_dir": os.path.join(rootdir, 'patches'),
+        "compress_rom": "Patch", 
+        "enable_distribution_file": "True",
+        "distribution_file": os.path.join(rootdir, "data", "random_settings.json"),
+        "create_spoiler": "True",
+        "world_count": worldcount
+    }
+
+    with open(os.path.join('data', 'randomizer_settings.json'), 'w') as fp:
+        json.dump(randomizer_settings, fp, indent=4)
+
+
+def generate_patch_file(max_retries=3):
+    retries = 0
+    while(True):
+        status_code = subprocess.call(
+            ["python3", os.path.join("randomizer", "OoTRandomizer.py"), "--settings", os.path.join("..", "data", "randomizer_settings.json")],
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL)
```

Comment:
I would prefer the output to be convinced the script is actually doing something.

Reply:
The output also points out where the resulting files are placed.

Reply:
Will the GUI disable terminal output? If not, hiding randomizer output like this is a good idea to avoid spoilers.

Reply:
The concern is that the output includes the settings string and seed

Reply:
I just ran it once and see
```
You are on version 6.0.36 R-1, and the latest is version 6.0.39 R-1.
OoT Randomizer Version 6.0.36 R-1  -  Seed: AIWEANOJ7K
(Original) Settings string: AJFBFGCSKSF6J4TRABZE9AB2EEFCJBASAJADABNAESBQ2EJBN6DLED5JH8TSNEF3KUUXEAAKAAMSPDMNEC

Generating World 1.
Creating Overworld
Calculating Access Rules.
Generating Item Pool.
Setting Entrances.
Failed to place all entrances in a pool for world 0. Will retry 19 more times
        No more valid entrances to replace with Lon Lon Ranch -> LLR Tower in world 0
Failed to place all entrances in a pool for world 0. Will retry 18 more times
        No more valid entrances to replace with Market -> Market Bombchu Bowling in world 0
Failed to place all entrances in a pool for world 0. Will retry 17 more times
        No more valid entrances to replace with Market -> Market Treasure Chest Game in world 0
Failed to place all entrances in a pool for world 0. Will retry 16 more times
        No more valid entrances to replace with Kak Backyard -> Kak Odd Medicine Building in world 0
Failed to place all entrances in a pool for world 0. Will retry 15 more times
        No more valid entrances to replace with Kak Backyard -> Kak Odd Medicine Building in world 0
Failed to place all entrances in a pool for world 0. Will retry 14 more times
        No more valid entrances to replace with GV Fortress Side -> GV Carpenter Tent in world 0
Failed to place all entrances in a pool for world 0. Will retry 13 more times
        No more valid entrances to replace with Zoras Domain -> ZD Shop in world 0
Failed to place all entrances in a pool for world 0. Will retry 12 more times
        No more valid entrances to replace with Lon Lon Ranch -> LLR Stables in world 0
Failed to place all entrances in a pool for world 0. Will retry 11 more times
        No more valid entrances to replace with Goron City -> GC Shop in world 0
Failed to place all entrances in a pool for world 0. Will retry 10 more times
        No more valid entrances to replace with DMC Lower Nearby -> DMC Great Fairy Fountain in world 0
Failed to place all entrances in a pool for world 0. Will retry 9 more times
        No more valid entrances to replace with LH Fishing Island -> LH Fishing Hole in world 0
Failed to place all entrances in a pool for world 0. Will retry 8 more times
        No more valid entrances to replace with Graveyard -> Graveyard Dampes House in world 0
Failed to place all entrances in a pool for world 0. Will retry 7 more times
        No more valid entrances to replace with Market Back Alley -> Market Bombchu Shop in world 0
Failed to place all entrances in a pool for world 0. Will retry 6 more times
        No more valid entrances to replace with Kakariko Village -> Kak Bazaar in world 0
Failed to place all entrances in a pool for world 0. Will retry 5 more times
        No more valid entrances to replace with Market -> Market Shooting Gallery in world 0
Failed to place all entrances in a pool for world 0. Will retry 4 more times
        No more valid entrances to replace with Market -> Market Shooting Gallery in world 0
Failed to place all entrances in a pool for world 0. Will retry 3 more times
        No more valid entrances to replace with Market Entrance -> Market Guard House in world 0
Failed to place all entrances in a pool for world 0. Will retry 2 more times
        No more valid entrances to replace with Lon Lon Ranch -> LLR Tower in world 0
Failed to place all entrances in a pool for world 0. Will retry 1 more times
        No more valid entrances to replace with Market -> Market Bazaar in world 0
Failed to place all entrances in a pool for world 0. Will retry 0 more times
        No more valid entrances to replace with Kak Impas Ledge -> Kak Impas House Back in world 0
Failed attempt 1 of 10: Entrance placement attempt count exceeded for world 0
Retrying...


Generating World 1.
Creating Overworld
Calculating Access Rules.
Generating Item Pool.
Setting Entrances.
Fill the world.
Placing shop items.
Placed shop items for world 1.
Placing dungeon items.
Placing song items.
Placed song items for world 1.
Placing progression items.
Placing priority items.
Placing the rest of the items.
Calculating playthrough.
Collected 18 spheres
Collected 17 final spheres
Calculating hint data.
Hint copies is zero for type trial. Assuming this hint type should be disabled.
Hint copies is zero for type always. Assuming this hint type should be disabled.
Hint copies is zero for type woth. Assuming this hint type should be disabled.
Hint copies is zero for type barren. Assuming this hint type should be disabled.
Hint copies is zero for type entrance. Assuming this hint type should be disabled.
Hint copies is zero for type sometimes. Assuming this hint type should be disabled.
Hint copies is zero for type random. Assuming this hint type should be disabled.
Hint copies is zero for type item. Assuming this hint type should be disabled.
Hint copies is zero for type song. Assuming this hint type should be disabled.
Hint copies is zero for type overworld. Assuming this hint type should be disabled.
Hint copies is zero for type dungeon. Assuming this hint type should be disabled.
Hint copies is zero for type named-item. Assuming this hint type should be disabled.
Patching ROM.
Created patchfile at: /home/matthewkirby/plando-random-settings/patches/OoT_FB35F_AIWEANOJ7K.zpf
Created spoiler log at: OoT_FB35F_AIWEANOJ7K_Spoiler.json
Copied distribution file to: /home/matthewkirby/plando-random-settings/patches/OoT_FB35F_AIWEANOJ7K_Distribution.json
Done. Enjoy.
```

So you know settings string, seed, indoor er from the error messages, sphere count, songsanity

Reply:
Also, for whatever reason, the entirety of the text is being sent to `stderr`.

Reply:
Solution was to just print our own output
```
RSL GENERATOR: RUNNING THE RANDOMIZER - ATTEMPT 1 OF 3
RSL GENERATOR: RUNNING THE RANDOMIZER - ATTEMPT 2 OF 3
RSL GENERATOR: RUNNING THE RANDOMIZER - ATTEMPT 3 OF 3
RSL GENERATOR ERROR: MAX RETRIES (3) REACHED. RESELECTING SETTINGS.
RSL GENERATOR: RUNNING THE RANDOMIZER - ATTEMPT 1 OF 3
RSL GENERATOR: PATCH FILE SAVED IN: /home/matthewkirby/plando-random-settings/patches
```